File: lib/banner_ad_view.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'ad.dart';

/// Widget for banner ad
class BannerAdView extends StatelessWidget {
  final String unitId;
  final double width;
  final double height;
  final AdCallback onAdLoad;
  final AdCallback onAdShow;
  final AdCallback onAdClick;
  final AdCallback onAdClose;
  final AdErrorCallback onError;

  BannerAdView(
      {Key key,
      this.unitId,
      this.width,
      this.height,
      this.onAdLoad,
      this.onAdShow,
      this.onAdClick,
      this.onAdClose,
      this.onError})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    Widget banner;
    if (defaultTargetPlatform == TargetPlatform.android) {
      banner = AndroidView(
        viewType: 'com.mob.adsdk/banner',
        creationParams: {
          "unitId": unitId,
          "width": width,
          "height": height,
        },
        creationParamsCodec: const StandardMessageCodec(),
        onPlatformViewCreated: _onPlatformViewCreated,
      );
    } else if (defaultTargetPlatform == TargetPlatform.iOS) {
      banner = UiKitView(
        viewType: 'com.mob.adsdk/banner',
        creationParams: {
          "unitId": unitId,
          "width": width,
          "height": height,
        },
        creationParamsCodec: const StandardMessageCodec(),
        onPlatformViewCreated: _onPlatformViewCreated,
      );
    } else {
      banner = Text("Not supported");
    }

    return Container(
      width: width,
      height: height,
      child: banner,
    );
  }

  void _onPlatformViewCreated(int id) {
    EventChannel eventChannel = EventChannel("com.mob.adsdk/banner_event_$id");
    eventChannel.receiveBroadcastStream().listen((event) {
      switch (event["event"]) {
        case "onAdLoad":
          onAdLoad?.call(event["id"]);
          break;

        case "onAdShow":
          onAdShow?.call(event["id"]);
          break;

        case "onAdClick":
          onAdClick?.call(event["id"]);
          break;

        case "onAdClose":
          onAdClose?.call(event["id"]);
          break;

        case "onError":
          onError?.call(event["id"], event["code"], event["message"]);
          break;
      }
    });
  }
}

File: android/src/main/java/com/mob/adsdk/flutterplugin/BannerAdView.java
package com.mob.adsdk.flutterplugin;

import android.app.Activity;
import android.content.Context;
import android.view.View;
import android.widget.RelativeLayout;

import com.mob.adsdk.AdSdk;

import io.flutter.app.FlutterApplication;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.platform.PlatformView;

import java.util.HashMap;
import java.util.Map;

public class BannerAdView implements PlatformView {

    private EventChannel.EventSink mEventSink;
    private RelativeLayout mContainer;
    private AdSdk.BannerAd mBannerAd;

    public BannerAdView(Context context, BinaryMessenger binaryMessenger, int id, String unitId, float width, float height) {
        EventChannel eventChannel = new EventChannel(binaryMessenger, "com.mob.adsdk/banner_event_" + id);
        eventChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object o, EventChannel.EventSink eventSink) {
                mEventSink = eventSink;
            }

            @Override
            public void onCancel(Object o) {
            }
        });

        Activity activity = ((FlutterApplication) context.getApplicationContext()).getCurrentActivity();
        mContainer = new RelativeLayout(context);
        AdSdk.getInstance().loadBannerAd(activity, unitId, mContainer, width, height,
                new AdSdk.BannerAdListener() {
                    @Override
                    public void onAdLoad(String id, AdSdk.BannerAd ad) {
                        // 设置刷新频率，为0或30~120之间的整数，单位秒，0表示不自动轮播，默认30秒
                        ad.setRefreshInterval(30);
                        mBannerAd = ad;

                        Map<String, Object> result = new HashMap<>();
                        result.put("id", id);
                        result.put("event", "onAdLoad");
                        mEventSink.success(result);
                    }

                    @Override
                    public void onAdShow(String id) {
                        Map<String, Object> result = new HashMap<>();
                        result.put("id", id);
                        result.put("event", "onAdShow");
                        mEventSink.success(result);
                    }

                    @Override
                    public void onAdClose(String id) {
                        Map<String, Object> result = new HashMap<>();
                        result.put("id", id);
                        result.put("event", "onAdClose");
                        mEventSink.success(result);
                        mEventSink.endOfStream();
                    }

                    @Override
                    public void onAdClick(String id) {
                        Map<String, Object> result = new HashMap<>();
                        result.put("id", id);
                        result.put("event", "onAdClick");
                        mEventSink.success(result);
                    }

                    @Override
                    public void onError(String id, int code, String message) {
                        Map<String, Object> result = new HashMap<>();
                        result.put("id", id);
                        result.put("event", "onError");
                        result.put("code", code);
                        result.put("message", message);
                        mEventSink.success(result);
                        mEventSink.endOfStream();
                    }
                });
    }

    @Override
    public View getView() {
        return mContainer;
    }

    @Override
    public void dispose() {
        if (null != mBannerAd) {
            mBannerAd.destroy();
            mBannerAd = null;
        }
    }
}
